### PR TITLE
PCHR-2457: Document open buttons take very long to load on ssp

### PR DESCRIPTION
**Pull request by bmnepali**
## Overview
This PR makes necessary changes in T&A document modal markup and modal controller to make the spinner load in document modal with all correct data in SSP.

**Note:** 
Before and after sections are describes in https://github.com/compucorp/civihr-employee-portal/pull/358

## Technical Details
1. Add the `<crm-loading></crm-loading>` directive in document modal markup:
```html
<crm-loading show="!isLoading || false">
  <div class="modal-body" nv-file-drop uploader="documentModal.uploader">
  </div>
  <div class="modal-footer">
  </div>
</crm-loading>
```

2. Use "common/directives/loading" in run.module.js in T&A.
```js
define[
   ...
   common/directives/loading,
   ...
   function () { 
   ...
   }]
```

3. Separate the code to collect contacts in a new function. The need to separate is that we have to frequently update the contacts bases on the changes in cached contacts array.
 
```js
function collectContacts () {
   vm.contacts = {
        target: initialContacts('target'),
        assignee: initialContacts('assignee')
   };
}
```
  
4. Watch for the change in contact search array and update the contacts when there is a change in cached contacts. The need to watch the cached contacts is that, when a document modal is opened form SSP, the modal is loaded before the contacts are loaded and cached which makes the collected contacts not up-to-date. So, when the contacts are loaded and cached we are re-collecting the contacts in T&A document modal controller using `collectContacts()` function.

```js
function initWatches () {
   $rootScope.$watch('cache.contact.arrSearch', collectContacts);
}
```

5. Call function to college and watch contact search array in controller initialize. Also define "isLoading" variable in "$rootscope" for check if the state is in loading anything in background.

```js
(function init () {
    ...
    collectContacts();
    initWatches();
})();
```

- [x] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.tasksassignments/views/modal/document.html [hunk 1]
```diff
@@ -9,7 +10,8 @@ <h4 class="modal-title" ng-switch="documentModal.isRole('admin')">
       <span ng-switch-when="false">{{ documentModal.getDocumentType(documentModal.document.activity_type_id) || 'Select Document Type' }}</span>
     </h4>
   </div>
-  <div class="modal-body" nv-file-drop uploader="documentModal.uploader">
+  <crm-loading ng-if="isLoading"></crm-loading>
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.tasksassignments/views/modal/document.html]: the `<crm-loading>` directive should *wrap* the content that it's supposed to hide. And also it has a `show` parameter that you should use to decide when to hide the spinner.
```html
<crm-loading show="!isLoading">
  <div class="modal-body" nv-file-drop uploader="documentModal.uploader">
  </div>
  <div class="modal-footer">
  </div>
</crm-loading>
```

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js [hunk 2]
```diff
@@ -71,8 +71,10 @@ define([
     vm.viewFile = viewFile;
 
     (function init () {
+      $rootScope.isLoading = false;
       angular.copy(data, vm.document);
       angular.copy(files, vm.files);
+
       vm.data = data;
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js]: please put all this assignments in a separate functions (or multiple functions, depending if the assignments are related or not), this `init` function is too big

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js [hunk 3]
```diff
@@ -549,5 +561,16 @@ define([
         $rootScope.$broadcast('ct-spinner-hide');
       });
     }
+
+    /**
+     * Executes watchers for changes
+     */
+    function watchForChanges () {
+      // Watches the changes in cached contacts search array
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js]: unnecessary comment, the code is self explanatory

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js [hunk 4]
```diff
@@ -549,5 +561,16 @@ define([
...
+      $rootScope.$watch('cache.contact.arrSearch', function (newValue) {
+        // Update the collected contacts
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js]: unnecessary comment, the code is self explanatory

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js [hunk 5]
```diff
@@ -549,5 +561,16 @@ define([
         $rootScope.$broadcast('ct-spinner-hide');
       });
     }
+
+    /**
+     * Executes watchers for changes
+     */
+    function watchForChanges () {
+      // Watches the changes in cached contacts search array
+      $rootScope.$watch('cache.contact.arrSearch', function (newValue) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js]: if you don't need to use `newValue` and/or `oldValue`, you can write this as
```js
 $rootScope.$watch('cache.contact.arrSearch', collectContacts);
```

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js [hunk 6]
```diff
@@ -549,5 +561,16 @@ define([
         $rootScope.$broadcast('ct-spinner-hide');
       });
     }
+
+    /**
+     * Executes watchers for changes
+     */
+    function watchForChanges () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js]: call this `initWatchers`

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js (hunk 3 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js]: Ok! I will remove it!

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js (hunk 4 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js]: ok! will remove it!

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js (hunk 2 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js]: Done!

**Pull request merged**